src/dachshund/candidate.rs

```diff
@@ -387,14 +396,20 @@ where
         Ok(())
     }
 
-    /// Create a copy of itself, needed for expand_with_node.
-    /// This happens for every candidate we want to score, not just
-    /// the ones we plan on expanding, so the performance of the
-    /// search is very sensitive to the cost of this operation.
+    /// Create a copy of itself, needed for initializing candidate
+    /// from node. This happens for every candidate we want to add to
+    /// the beam for the next epoch. So the performance of the
+    /// search is sensitive to the cost of this operation.
     pub fn replicate(&self, keep_score: bool) -> Self {
+        // We clone these hashmaps very frequently so
+        // we keep capacity artificially low.
+        let mut new_neighborhood = self.neighborhood.clone();
+        if 2 * new_neighborhood.capacity() > new_neighborhood.len() {
+           new_neighborhood.shrink_to_fit()
+        }
+
```

Comment:
Is this purely for memory-usage reasons? These feels like this was a response to a bug that was especially hard to figure out.

Reply:
This actually more about time than memory usage, though they're related here. In a normal use case, you want the hashmap to overallocate by quite a bit because it saves you from having to reallocate when you grow it too frequently. For our usecase, we do one clone of the hashmap for each time we adjust the neighborhood (which involves a number of updates equal to the degree of the vertex we're adding). 

Since we're clone the hashmaps very frequently so we effectively have to re-allocate all the time and the "smart" overallocation costs us quite a bit. On workloads corresponding to dense graphs, this allocation and deallocation could easily be ~40% of the rumtime before this optimization.